primo/utils/map_utils.py

```diff
@@ -22,548 +21,352 @@
+
+        well_type_to_plot : str, optional
+            Type of well to plot ('Gas' or 'Oil'). Default is None.
+
+        campaign : Campaign, optional
+            Campaign instance to map wells to projects when visualize_type is 'project'.
+            Default is None.
+
+        Raises
+        ------
+        ValueError
+            If visualize_type is 'project' but campaign is not provided.
+        """
+
+        # Inline logic for generating cluster or project colors
+        colors = [
+            "red",
+            "blue",
+            "green",
+            "orange",
+            "purple",
+            "yellow",
+            "cyan",
+            "magenta",
+            "pink",
+            "brown",
+            "black",
+        ]
+
+        # Marker logic for project visualization
+        if visualize_type == "project":
+            if campaign is None:
+                raise ValueError(
+                    "A Campaign instance must be provided when visualize_type is 'project'."
+                )
+            project_ids = set()  # Use a set to avoid duplicates
+            for row in self.df.itertuples():
+                well_id = row[
+                    self.df.columns.get_loc(self.well_data.col_names.well_id) + 1
+                ]
+                project_id = campaign.get_project_id_by_well_id(well_id)
+                if project_id is not None:
+                    project_ids.add(project_id)
+
+            # Generate color mapping for project IDs
+            project_colors = {
+                project_id: colors[i % len(colors)]
+                for i, project_id in enumerate(project_ids)
+            }
+
+            for row in self.df.itertuples():
+                well_id = row[
+                    self.df.columns.get_loc(self.well_data.col_names.well_id) + 1
+                ]
+                project_id = campaign.get_project_id_by_well_id(well_id)
+
+                if project_id is not None:
+                    popup_text = f"Candidate Project: Project {project_id}"
+                    folium.CircleMarker(
+                        location=[row.geometry.y, row.geometry.x],
+                        radius=5,
+                        popup=popup_text,
+                        fill=True,
+                        color=project_colors.get(project_id, "gray"),
+                    ).add_to(map_obj)
+
+        # Marker logic for well visualization
+        else:
+            # Iterate over the rows using itertuples and access columns by index
+            for row in self.df.itertuples():
+                well_id = row[
+                    self.df.columns.get_loc(self.well_data.col_names.well_id) + 1
+                ]
+                age = row[self.df.columns.get_loc(self.well_data.col_names.age) + 1]
+                depth = row[self.df.columns.get_loc(self.well_data.col_names.depth) + 1]
+
+                popup_text = f"API: {well_id}<br>Age: {age}<br>Depth: {depth}"
```

Comment:
We allow users to use any well_id system. Could we make the "API:" section an argument? For example, 
base_template = f"{well_id}<br>Age: {age}<br>Depth: {depth}"
popup_text = well_id_type + base_template. 
Where well_id_type is the argument and f"API: " is the default value.

Reply:
Replacing 'API' with 'Well ID' to make it more general.